src/memray/_memray/record_writer.cpp

```diff
@@ -93,23 +392,266 @@ RecordWriter::writeTrailer()
+        return false;
+    }
+
+    return true;
 }
 
 std::unique_ptr<RecordWriter>
-RecordWriter::cloneInChildProcess()
+AggregatingRecordWriter::cloneInChildProcess()
 {
     std::unique_ptr<io::Sink> new_sink = d_sink->cloneInChildProcess();
     if (!new_sink) {
         return {};
     }
-    return std::make_unique<RecordWriter>(
+    return std::make_unique<AggregatingRecordWriter>(
             std::move(new_sink),
             d_header.command_line,
             d_header.native_traces);
 }
 
+bool
+AggregatingRecordWriter::writeRecord(const MemoryRecord& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    MemorySnapshot snapshot;
+    snapshot.ms_since_epoch = record.ms_since_epoch;
+    snapshot.rss = record.rss;
+    snapshot.heap = d_high_water_mark_aggregator.getCurrentHeapSize();
+    d_memory_snapshots.push_back(snapshot);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeRecord(const pyrawframe_map_val_t& item)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    d_stats.n_frames += 1;
+    const auto& [frame_id, raw] = item;
+    d_frames_by_id.emplace(
+            frame_id,
+            Frame{raw.function_name, raw.filename, raw.lineno, raw.is_entry_frame});
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeRecord(const UnresolvedNativeFrame& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    d_native_frames.emplace_back(record);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeMappings(const std::vector<ImageSegments>& mappings)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    d_mappings_by_generation.push_back(mappings);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeThreadSpecificRecord(thread_id_t tid, const FramePop& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    auto count = record.count;
+    auto& stack = d_python_stack_ids_by_thread[tid];
+    assert(stack.size() >= record.count);
+    while (count) {
+        count -= 1;
+        stack.pop_back();
+    }
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeThreadSpecificRecord(thread_id_t tid, const FramePush& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    auto [it, inserted] = d_python_stack_ids_by_thread.emplace(tid, python_stack_ids_t{});
+    auto& stack = it->second;
+    if (inserted) {
+        stack.reserve(1024);
+    }
+    FrameTree::index_t current_stack_id = stack.empty() ? 0 : stack.back();
+    FrameTree::index_t new_stack_id =
+            d_python_frame_tree.getTraceIndex(current_stack_id, record.frame_id);
+    stack.push_back(new_stack_id);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeThreadSpecificRecord(thread_id_t tid, const AllocationRecord& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    Allocation allocation;
+    allocation.tid = tid;
+    allocation.address = record.address;
+    allocation.size = record.size;
+    allocation.allocator = record.allocator;
+    allocation.native_frame_id = 0;
```

Comment:
Any reason why we are not just doing

```suggestion
    Allocation allocation{
    tid,
    record.address,
    record.size,
    record.allocator,
    }
```

?